compiler/fir/checkers/checkers.jvm/src/org/jetbrains/kotlin/fir/analysis/jvm/checkers/expression/FirJavaGenericVarianceViolationTypeChecker.kt

```diff
@@ -0,0 +1,99 @@
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirDeclarationOrigin
+import org.jetbrains.kotlin.fir.declarations.FirFunction
+import org.jetbrains.kotlin.fir.expressions.FirFunctionCall
+import org.jetbrains.kotlin.fir.expressions.impl.FirArgumentListWithResolutionInfo
+import org.jetbrains.kotlin.fir.expressions.toResolvedCallableSymbol
+import org.jetbrains.kotlin.fir.resolve.inference.inferenceComponents
+import org.jetbrains.kotlin.fir.types.*
+import org.jetbrains.kotlin.types.AbstractTypeChecker
+import org.jetbrains.kotlin.types.model.TypeConstructorMarker
+import org.jetbrains.kotlin.types.model.typeConstructor
+
+/**
+ * Checks compatibility of variance of type argument for Java collections.
+ *
+ * Java collection interfaces all include methods mutating the collection. Hence, they naturally map to mutable versions of Kotlin
+ * collections. But Kotlin provides immutable collections and it's good practice to enforce immutability when possible. Hence, to make it
+ * easier to use immutable collections in Kotlin, Java collection types are instead mapped to flexible type with mutable collection as the
+ * lower type and immutable collection as upper type. However, flexible types make type checking unsound. Hence we have this check that
+ * catches a common mistake allowed by flexible type.
+ *
+ * Consider a Java method accepting type `List<Object>`, which is mapped to `MutableList<Any?>..List<(out) Any?>`. If one passes a mutable
+ * list of some more concrete type than `Any` (ArrayList<String>, LinkedList<Int>, etc.) to this Java method, then any writes to this
+ * mutable collection could cause `ClassCastException`s if the same mutable list is read elsewhere. It's the purpose of this checker to
+ * reject such code.
+ *
+ * On the other hand, if one passes an immutable list of some more concrete type, then any writes to it on the Java side would cause
+ * `UnsupportedOperationException`, which is expected and the price we pay in order to make immutable collection easier to use. This checker
+ * doesn't do anything to prevent this from happening.
+ */
+object FirJavaGenericVarianceViolationTypeChecker : FirFunctionCallChecker() {
+    private val javaOrigin = setOf(FirDeclarationOrigin.Java, FirDeclarationOrigin.Enhancement)
+
+    override fun check(expression: FirFunctionCall, context: CheckerContext, reporter: DiagnosticReporter) {
+        val calleeFunction = expression.calleeReference.toResolvedCallableSymbol()?.fir as? FirFunction<*> ?: return
+        if (calleeFunction.origin !in javaOrigin) {
+            return
+        }
+        val argList = expression.argumentList as? FirArgumentListWithResolutionInfo
+        val typeParameterSubstitutor = argList?.typeParameterSubstitutor ?: return
```

Comment:
I'm not sure there's a really need on storing `typeParameterSubstitutor`
We may just create a new one based on `symbol.typeParameters => typeArguments` mapping

Reply:
IIUC there can be type variables declared in the environment (containing classes) right?

Reply:
I mean that having a field for `typeParameterSubstitutor` in `argList` seems to be redundant since you can build the substitutor via information already stored at `expression: FirFunctionCall` (one may build a map from `symbol.typeParameters` to the relevant `expression.typeArguments` and then pass it to`substitutorByMap`)

Reply:
Thanks! I misunderstood how type variables in containing classes work. Apparently in that case a synthetic symbol is created with these types substituted already. I will create the subsitutuor for type arguments instead.